include/hobbes/eval/cc.H

```diff
@@ -108,11 +95,13 @@ public:
   SearchEntries search(const std::string&, const MonoTypePtr&);
   SearchEntries search(const std::string&, const std::string&);
 private:
+  static std::recursive_mutex hccmtx;
```

Comment:
do _not_ use a statically initialized mutex

Reply:
You could define and use helpers such as these below:
```c++
static auto lock() -> std::recursive_mutex& {
  static std::recursive_mutex h;
  return h;
}
template<class Fn>
static auto hlock(Fn&& fn) -> decltype(fn()) {
  std::lock_guard<std::recursive_mutex> l{lock()};
  return fn();
}
```

Reply:
This helper idiom is much better than the original `hlock` design and using vanilla lock guard everywhere

Just one question, why _do not_ use statically initialized mutex? `h` in `lock()` is statically initialized as well, I don't know the difference

Reply:
Static initialization breaks linker dependency analysis, especially when it comes to lining against static libraries. This [blog post](https://www.bfilipek.com/2018/02/static-vars-static-lib.html) explains a little how tricky they are to get right.

Reply:
You are right

First of all, I got the life time of a static class member wrong, I always thought it is like local static, turns out it is just like global static

```cpp
#include <cstdio>

struct X {
    X() { std::puts("X()"); }
};

struct S {
    static X x;
};

X S::x{};

int main() {}
```
[It prints `X()` without initialization `S`](https://godbolt.org/z/n4qonhWPr), I didn't know this

Secondly, as always, C++ surprises me on daily basis. Let's do it your way, the correct way :)